### PR TITLE
[4.0] Form fields prepared statements, Part 2

### PR DESCRIPTION
**Pull request by SharkyKZ**
Pull Request for Issue # .

### Summary of Changes

Adds prepared statements and cleans up queries in some form fields.

### Testing Instructions

Install testing sample data.

1) Go to article list. Inspect `Select Stage` field.
2) Edit an article. Inspect `Status` field.
3) Go to module list. Inspect `Select Menu Item` field.
3) Edit a module. Inspect `Layout` field.
4) Edit a user. Inspect `Backend Template Style` and `Editor` fields.
5) Go to menu items list. Inspect `Select Menu` field.
6) Edit and save a contact, banner or newsfeed.

### Expected result

Works like before.

### Documentation Changes Required

No.

### CODE REVIEW

File: libraries/src/Form/Field/TemplatestyleField.php [hunk 1]
```diff
@@ -144,11 +144,23 @@ protected function getGroups()
...
+			[
+				$db->quoteName('s.id'),
+				$db->quoteName('s.title'),
+				$db->quoteName('e.name'),
+				$db->quoteName('s.template'),
+			]
+		)
+			->from($db->quoteName('#__template_styles', 's'))
+			->where($db->quoteName('s.client_id') . ' = :clientId')
+			->bind(':clientId', $client->id)
```
Comment (Quy) [inline libraries/src/Form/Field/TemplatestyleField.php]:
```suggestion
			->bind(':clientId', $client->id, ParameterType::INTEGER)
```

Comment (Quy) [reply libraries/src/Form/Field/TemplatestyleField.php]: Add `use Joomla\Database\ParameterType;`

File: libraries/src/Form/Field/OrderingField.php [hunk 2]
```diff
@@ -176,8 +177,9 @@ protected function getQuery()
 		$query = $db->getQuery(true);
 		$query->select(array($db->quoteName($ordering, 'value'), $db->quoteName($title, 'text')))
```
Comment (Quy) [inline libraries/src/Form/Field/OrderingField.php]:
```suggestion
		$query->select([$db->quoteName($ordering, 'value'), $db->quoteName($title, 'text')])
```

File: libraries/src/Form/Field/TransitionField.php [hunk 3]
```diff
@@ -150,8 +167,7 @@ function ($item) use ($user, $extension)
 		}
 
 		// Get workflow stage title
-		$query
-			->clear()
+		$query = $db->getQuery(true)
 			->select($db->quoteName('title'))
 			->from($db->quoteName('#__workflow_stages'))
 			->where($db->quoteName('id') . ' = ' . (int) $workflowStage);
```
Comment (Quy) [inline libraries/src/Form/Field/TransitionField.php]: Bind.

File: libraries/src/Form/Field/TransitionField.php [hunk 4]
```diff
@@ -150,11 +167,11 @@ function ($item) use ($user, $extension)
...
 
 		// Get workflow stage title
-		$query
-			->clear()
+		$query = $db->getQuery(true)
 			->select($db->quoteName('title'))
 			->from($db->quoteName('#__workflow_stages'))
-			->where($db->quoteName('id') . ' = ' . (int) $workflowStage);
+			->where($db->quoteName('id') . ' = :stage')
+			->bind(':stage', $workflowStage);
```
Comment (Quy) [inline libraries/src/Form/Field/TransitionField.php]:
```suggestion
			->bind(':stage', $workflowStage, ParameterType::INTEGER);
```

File: libraries/src/Form/Field/MenuField.php [hunk 5]
```diff
@@ -47,13 +48,27 @@ protected function getGroups()
...
+					$db->quoteName('client_id'),
+					$db->quoteName('title'),
+				]
+			);
 
 		if (strlen($clientId))
 		{
-			$query->where('client_id = ' . (int) $clientId);
+			$client = (int) $clientId;
+			$query->where('client_id = :client')
```
Comment (Quy) [inline libraries/src/Form/Field/MenuField.php]:
```suggestion
			$query->where($db->quoteName('client_id') . ' = :client')
```

File: libraries/src/Form/Field/TransitionField.php [hunk 6]
```diff
@@ -104,20 +105,36 @@ protected function getOptions()
...
+				]
+			)
+			->whereIn($db->quoteName('t.from_stage_id'), [-1, $workflowStage])
+			->where(
+				[
+					$db->quoteName('t.to_stage_id') . ' = ' . $db->quoteName('s.id'),
+					$db->quoteName('t.to_stage_id') . ' != :stage1',
+					$db->quoteName('s.workflow_id') . ' = ' . $db->quoteName('s2.workflow_id'),
+					$db->quoteName('s2.id') . ' = :stage2',
+					$db->quoteName('t.published') . '= 1',
```
Comment (Quy) [inline libraries/src/Form/Field/TransitionField.php]:
```suggestion
					$db->quoteName('t.published') . ' = 1',
```

File: libraries/src/Form/Field/TransitionField.php [hunk 7]
```diff
@@ -104,20 +105,36 @@ protected function getOptions()
...
+					$db->quoteName('s.published') . '= 1',
```
Comment (Quy) [inline libraries/src/Form/Field/TransitionField.php]:
```suggestion
					$db->quoteName('s.published') . ' = 1',
```

File: libraries/src/Form/Field/PluginsField.php [hunk 8]
```diff
@@ -123,16 +123,30 @@ protected function getOptions()
...
+						$db->quoteName('ordering'),
+						$db->quoteName('name'),
+					]
+				);
 
 			if ((string) $this->element['useaccess'] === 'true')
 			{
-				$groups = implode(',', Factory::getUser()->getAuthorisedViewLevels());
-				$query->where($db->quoteName('access') . ' IN (' . $groups . ')');
+				$query->whereIn($db->quoteName('access'), Factory::getUser()->getAuthorisedViewLevels());
```
Comment (alikon) [inline libraries/src/Form/Field/PluginsField.php:149]: i've always run into issues having the 2nd argument not as a variable , maybe is not the case with `whereIn()` ?

Comment (SharkyKZ) [reply libraries/src/Form/Field/PluginsField.php:149]: Yes, not a case with `whereIn()`. Only `bind()` requires value to be passed by reference.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 782ae7a912d5c7d218c9efae3054886d8af1702c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26069">issues.joomla.org/tracker/joomla-cms/26069</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on f3bbbfd83438336779d6561d02c125f1224689b1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26069">issues.joomla.org/tracker/joomla-cms/26069</a>.</sub>

**Comment by jwaisner** [comment]
@Quy Please test again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26069">issues.joomla.org/tracker/joomla-cms/26069</a>.</sub>

**Comment by SharkyKZ** [comment]
RTC as @Quy's test still good.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26069">issues.joomla.org/tracker/joomla-cms/26069</a>.</sub>

**Comment by HLeithner** [comment]
Thanks

**Pull request merged**